// nwb/lib/createProject.js
'use strict';

exports.__esModule = true;

var _PROJECT_CREATORS;

exports.getWebModulePrefs = getWebModulePrefs;
exports.npmModuleVars = npmModuleVars;
exports.validateProjectType = validateProjectType;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _copyTemplateDir = require('copy-template-dir');

var _copyTemplateDir2 = _interopRequireDefault(_copyTemplateDir);

var _inquirer = require('inquirer');

var _inquirer2 = _interopRequireDefault(_inquirer);

var _colours = require('./colours');

var _constants = require('./constants');

var _errors = require('./errors');

var _packageJson = require('../package.json');

var _packageJson2 = _interopRequireDefault(_packageJson);

var _utils = require('./utils');

var nwbVersion = _packageJson2['default'].version.split('.').slice(0, 2).concat('x').join('.');

function getWebModulePrefs(args, done) {
  // Determine defaults based on arguments
  var umd = true;
  if (args.umd === false) {
    umd = false;
  } else if (args.g || args.global) {
    umd = true;
  } else if (args.f || args.force) {
    umd = false;
  }
  var globalVariable = args.g || args.global || '';
  var jsNext = true;
  if (args.jsnext === false) {
    jsNext = false;
  }

  if (args.f || args.force) {
    return done(null, { umd: umd, globalVariable: globalVariable, jsNext: jsNext });
  }

  _inquirer2['default'].prompt([{
    type: 'confirm',
    name: 'umd',
    message: 'Do you want to create a UMD build for npm?',
    'default': umd
  }, {
    when: function when(_ref) {
      var umd = _ref.umd;
      return umd;
    },
    type: 'input',
    name: 'globalVariable',
    message: 'Which global variable should the UMD build export?',
    'default': globalVariable
  }, {
    type: 'confirm',
    name: 'jsNext',
    message: 'Do you want to create an ES6 modules build for npm?',
    'default': jsNext
  }]).then(function (answers) {
    return done(null, answers);
  }, function (err) {
    return done(err);
  });
}

function logCreatedFiles(targetDir, createdFiles) {
  createdFiles.sort().forEach(function (createdFile) {
    var relativePath = _path2['default'].relative(targetDir, createdFile);
    console.log('  ' + _colours.green('create') + ' ' + relativePath);
  });
}

function npmModuleVars(vars) {
  vars.jsNextMain = vars.jsNext ? '\n  "jsnext:main": "es6/index.js",' : '';
  return vars;
}

function validateProjectType(projectType) {
  if (!projectType) {
    throw new _errors.UserError('nwb: a project type must be provided, one of: ' + _constants.PROJECT_TYPES.join(', '));
  }
  if (_constants.PROJECT_TYPES.indexOf(projectType) === -1) {
    throw new _errors.UserError('nwb: project type must be one of: ' + _constants.PROJECT_TYPES.join(', '));
  }
}

var PROJECT_CREATORS = (_PROJECT_CREATORS = {}, _PROJECT_CREATORS[_constants.REACT_APP] = function (args, name, targetDir, cb) {
  var templateDir = _path2['default'].join(__dirname, '../templates/' + _constants.REACT_APP);
  var reactVersion = args.react || _constants.REACT_VERSION;
  var templateVars = { name: name, nwbVersion: nwbVersion, reactVersion: reactVersion };
  _copyTemplateDir2['default'](templateDir, targetDir, templateVars, function (err, createdFiles) {
    if (err) return cb(err);
    logCreatedFiles(targetDir, createdFiles);
    console.log('nwb: installing dependencies');
    try {
      _utils.installReact({ cwd: targetDir, version: reactVersion, save: true });
    } catch (e) {
      return cb(e);
    }
    cb();
  });
}, _PROJECT_CREATORS[_constants.REACT_COMPONENT] = function (args, name, targetDir, cb) {
  getWebModulePrefs(args, function (err, prefs) {
    if (err) return cb(err);
    var umd = prefs.umd;
    var globalVariable = prefs.globalVariable;
    var jsNext = prefs.jsNext;

    var templateDir = _path2['default'].join(__dirname, '../templates/' + _constants.REACT_COMPONENT);
    var reactVersion = args.react || _constants.REACT_VERSION;
    var templateVars = npmModuleVars({ umd: umd, globalVariable: globalVariable, jsNext: jsNext, name: name, nwbVersion: nwbVersion, reactVersion: reactVersion });
    _copyTemplateDir2['default'](templateDir, targetDir, templateVars, function (err, createdFiles) {
      if (err) return cb(err);
      logCreatedFiles(targetDir, createdFiles);
      console.log('nwb: installing dependencies');
      try {
        _utils.installReact({ cwd: targetDir, version: reactVersion, dev: true, save: true });
      } catch (e) {
        return cb(e);
      }
      cb();
    });
  });
}, _PROJECT_CREATORS[_constants.WEB_APP] = function (args, name, targetDir, cb) {
  var templateDir = _path2['default'].join(__dirname, '../templates/' + _constants.WEB_APP);
  var templateVars = { name: name, nwbVersion: nwbVersion };
  _copyTemplateDir2['default'](templateDir, targetDir, templateVars, function (err, createdFiles) {
    if (err) return cb(err);
    logCreatedFiles(targetDir, createdFiles);
    cb();
  });
}, _PROJECT_CREATORS[_constants.WEB_MODULE] = function (args, name, targetDir, cb) {
  getWebModulePrefs(args, function (err, prefs) {
    if (err) return cb(err);
    var umd = prefs.umd;
    var globalVariable = prefs.globalVariable;
    var jsNext = prefs.jsNext;

    var templateDir = _path2['default'].join(__dirname, '../templates/' + _constants.WEB_MODULE);
    var templateVars = npmModuleVars({ umd: umd, globalVariable: globalVariable, jsNext: jsNext, name: name, nwbVersion: nwbVersion });
    _copyTemplateDir2['default'](templateDir, targetDir, templateVars, function (err, createdFiles) {
      if (err) return cb(err);
      logCreatedFiles(targetDir, createdFiles);
      cb();
    });
  });
}, _PROJECT_CREATORS);

exports['default'] = function (args, type, name, dir, cb) {
  PROJECT_CREATORS[type](args, name, dir, cb);
};
