// nwb/lib/webpackBuild.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports['default'] = webpackBuild;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _argvSetEnv = require('argv-set-env');

var _argvSetEnv2 = _interopRequireDefault(_argvSetEnv);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _createWebpackConfig = require('./createWebpackConfig');

var _createWebpackConfig2 = _interopRequireDefault(_createWebpackConfig);

var _debug = require('./debug');

var _debug2 = _interopRequireDefault(_debug);

var _getPluginConfig = require('./getPluginConfig');

var _getPluginConfig2 = _interopRequireDefault(_getPluginConfig);

var _getUserConfig = require('./getUserConfig');

var _getUserConfig2 = _interopRequireDefault(_getUserConfig);

var _utils = require('./utils');

function webpackBuild(args, buildConfig, cb) {
  if (buildConfig === undefined) buildConfig = {};

  // Don't override environment if it's already set
  if (!process.env.NODE_ENV) {
    // Set cross-platform environment variables based on any --set-env-NAME
    // arguments passed to the command.
    _argvSetEnv2['default']();
    // Default environment for a build
    if (!process.env.NODE_ENV) {
      process.env.NODE_ENV = 'production';
    }
  }

  var userConfig = _getUserConfig2['default'](args);
  var pluginConfig = _getPluginConfig2['default']();
  if (typeof buildConfig == 'function') {
    buildConfig = buildConfig(args);
  }

  var webpackConfig = _createWebpackConfig2['default'](_extends({}, buildConfig, {
    server: false
  }), pluginConfig, userConfig.webpack);

  _debug2['default']('webpack config: %s', _utils.deepToString(webpackConfig));

  var compiler = _webpack2['default'](webpackConfig);
  compiler.run(function (err, stats) {
    if (err) return cb(err);
    console.log(stats.toString({
      children: false,
      chunks: false,
      colors: true,
      modules: false
    }));
    cb();
  });
}

module.exports = exports['default'];
