// nwb/lib/commands/build-web-app.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _appConfig = require('../appConfig');

var _utils = require('../utils');

var _webpackBuild = require('../webpackBuild');

var _webpackBuild2 = _interopRequireDefault(_webpackBuild);

exports['default'] = function (args, cb) {
  var entry = args._[1] || 'src/index.js';
  var dist = args._[2] || 'dist';

  require('./clean-app')({ _: ['clean-app', dist] });

  console.log('nwb: build-web-app');
  _utils.copyPublicDir('public', dist);
  _webpackBuild2['default'](args, {
    devtool: 'source-map',
    entry: {
      app: _path2['default'].resolve(entry)
    },
    output: {
      filename: '[name].js',
      path: _path2['default'].resolve(dist),
      publicPath: '/'
    },
    plugins: {
      html: _appConfig.getDefaultHTMLConfig(),
      vendorChunkName: 'vendor'
    }
  }, cb);
};

module.exports = exports['default'];
